src/core/Distribution.pm

```diff
@@ -1,38 +1,93 @@
-class Distribution {
-    has $.name;
-    has $.auth;
-    has $.author;
-    has $.authority;
-    has $.api;
-    has $.ver;
-    has $.version;
-    has $.description;
-    has @.depends;
-    has %.provides;
-    has %.files;
-    has $.source-url;
-    method auth { $!auth // $!author // $!authority }
-    method ver  { $!ver // $!version }
-    method hash {
-        {
-            :$!name,
-            :$.auth,
-            :$.ver,
-            :$!description,
-            :@!depends,
-            :%!provides,
-            :%!files,
-            :$!source-url,
-        }
+# API to obtain the data of any addressable content
+role Distribution {
+    # `meta` provides an API to the meta data in META6 spec (s22)
+    #   -   A Distribution may be represented internally by some other
+    #       spec (such as using the file system itself for prereqs), as
+    #       long as it can also be represented as the META6 hash format
+    method meta(--> Hash) { ... }
+
+    # `content($content-id)` provides an API to the data itself
+    #   -   Use `.meta` to determine the $address of a specific $content-id
+    #   -   IO::Handle is meant to be a data stream that may or may not be available; for now
+    #       it would return an IO::Handle and have `.open.slurp-rest(:bin)` called on it. So if
+    #       a socket wants to handle this role currently it would have to wrap `open` or `.slurp-rest`
+    #       to handle any protocol negotiation as well as probably saving the data to a tmpfile and
+    #       return an IO::Handle to that
+    method content($content-id --> IO::Handle) { ... }
+}
+
+role Distribution::Locally {
+    has IO::Path $.prefix;
+    method content($address) {
+        my $handle = IO::Handle.new: path => IO::Path.new($address, :CWD($!prefix // $*CWD));
+        $handle // $handle.throw;
+    }
+}
+
+# A distribution passed to `CURI.install()` will get encapsulated in this
+# class, which normalizes the meta6 data and adds identifiers/content-id
+class CompUnit::Repository::Distribution {
+    has Distribution $!dist handles 'content';
+    has $!meta;
+    submethod BUILD(:$!meta, :$!dist) { }
+    method new(Distribution $dist) {
+        my $meta = $dist.meta.hash;
+        $meta<ver>  //= $meta<version>;
+        $meta<auth> //= $meta<authority> // $meta<author>;
+        self.bless(:$dist, :$meta);
     }
+    method meta { $!meta }
+
     method Str() {
-        return "{$.name}:ver<{$.ver  // ''}>:auth<{$.auth // ''}>:api<{$.api // ''}>";
+        return "{$.meta<name>}"
+        ~ ":ver<{$.meta<ver>   // ''}>"
+        ~ ":auth<{$.meta<auth> // ''}>"
+        ~ ":api<{$.meta<api>   // ''}>";
+
     }
+
     method id() {
         return nqp::sha1(self.Str);
     }
 }
 
+class Distribution::Hash does Distribution does Distribution::Locally {
```

Comment:
Possibly; I went back and forth on it being a class and role because the amount of duplicated code seemed trivial. It just happens that I stopped bike shedding that detail while it was a `class`

See https://gist.github.com/ugexe/d9e0f8697b9c507e753a#file-distribution-pm6-L62 (from an older implementation) to preview such a change